src/Compilers/CSharp/Portable/Binder/PatternExplainer.cs

```diff
@@ -0,0 +1,512 @@
+                    // All we have are numeric constraints. Process them to compute a value not covered.
+                    var remainingValues = fac.AllValues;
+                    foreach (var constraint in constraints)
+                    {
+                        var (test, sense) = constraint;
+                        switch (test)
+                        {
+                            case BoundDagValueTest v:
+                                addRelation(BinaryOperatorKind.Equal, v.Value);
+                                break;
+                            case BoundDagRelationalTest r:
+                                addRelation(r.Relation, r.Value);
+                                break;
+                        }
+                        void addRelation(BinaryOperatorKind relation, ConstantValue value)
+                        {
+                            var filtered = fac.Related(relation, value);
+                            if (!sense)
+                                filtered = filtered.Complement();
+                            remainingValues = remainingValues.Intersect(filtered);
+                        }
+                    }
+
+                    if (remainingValues.Complement().IsEmpty)
+                        return "_";
+
+                    return SampleValueString(remainingValues, input.Type, requireExactType: requireExactType);
+                }
+
+                return null;
+            }
+
+            // Handle the special case of a recursive pattern
+            string tryHandleRecursivePattern()
+            {
+                if (constraints.IsEmpty && evaluations.IsEmpty)
+                    return null;
+
+                string typeName = requireExactType ? input.Type.ToDisplayString() : null;
+
+                foreach (var con in constraints)
+                {
+                    switch (con)
+                    {
+                        case (test: BoundDagNonNullTest _, sense: true):
+                        case (test: BoundDagExplicitNullTest _, sense: false):
+                            // not-null tests are implicitly incorporated into a recursive pattern
+                            break;
+                        default:
+                            return null;
+                    }
+                }
+
+                string deconstruction = null;
+                var properties = new Dictionary<Symbol, string>();
+                bool needsPropertyString = false;
+
+                foreach (var eval in evaluations)
+                {
+                    switch (eval)
+                    {
+                        case BoundDagDeconstructEvaluation e:
+                            var method = e.DeconstructMethod;
+                            int extensionExtra = method.RequiresInstanceReceiver ? 0 : 1;
+                            int count = method.Parameters.Length - extensionExtra;
+                            var subpatterns = new ArrayBuilder<string>(count);
```

Comment:
>var subpatterns = new ArrayBuilder<string>(count); [](start = 28, length = 50)

Minor: Consider using a `StringBuilder` instead and handling the "(..., ...)" formatting directly with the builder.